dartagnan/src/main/java/com/dat3m/dartagnan/verification/model/ExecutionModelManager.java

```diff
@@ -0,0 +1,74 @@
+package com.dat3m.dartagnan.verification.model;
+
+import com.dat3m.dartagnan.encoding.EncodingContext;
+import com.dat3m.dartagnan.program.memory.MemoryObject;
+import com.dat3m.dartagnan.verification.model.event.EventModelManager;
+import com.dat3m.dartagnan.verification.model.relation.RelationModelManager;
+import org.sosy_lab.java_smt.api.BooleanFormula;
+import org.sosy_lab.java_smt.api.Formula;
+import org.sosy_lab.java_smt.api.Model;
+import org.sosy_lab.common.configuration.InvalidConfigurationException;
+
+import java.math.BigInteger;
+import java.util.List;
+
+
+public class ExecutionModelManager {
+    private final EventModelManager eMManager;
+    private final RelationModelManager rMManager;
+
+    private EncodingContext context;
+    private Model model;
+    private ExecutionModelNext executionModel;
+
+    public ExecutionModelManager(){
+        eMManager = new EventModelManager(this);
+        rMManager = new RelationModelManager(this);
+    }
+
+    public ExecutionModelNext buildExecutionModel(EncodingContext context, Model model) {
+        if (this.context != null && this.model != null) {
+            if (this.context == context && this.model == model) {
+                return executionModel;
+            }
+        }
+
+        this.context = context;
+        this.model = model;
+        executionModel = new ExecutionModelNext(this);
+
+        eMManager.buildEventModels(executionModel, context);
+        extractMemoryLayout();
+        rMManager.buildRelationModels(executionModel, context);
+
+        return executionModel;
+    }
```

Comment:
The caching looks unnecessary IMO. Then it is also unnecessary to store `Model` and `Context`, meaning the class could be more or less stateless.
Especially storing `Model` can be problematic because `Model`s need to be explicitly closed to release resources (`model.close()`) which makes further accesses to them invalid.
Put differently, if you store the `Model`, then the lifetime of this class must be no longer than the lifetime of `Model`.

Reply:
OK, no caching then. But since the `Model` is indeed used during the whole process of constructing `ExecutionModelNext` and passing this object into all methods of three `Managers` makes code clumsy more or less, what about to store this object in the beginning of this method `buildExecutionModel()` and drop it at the last before return?

Reply:
I don't think code gets really clumsy if you pass one more parameter around. IMO, it often becomes even nicer code because more methods end up being fully self-contained, i.e., independent of everything else the class does.
Anyways, it's also fine to explicitly drop the model at the end.